File: src/value/iterator.zig
//! 迭代器值类型模块
//!
//! 定义 Glue 语言中用于遍历集合的迭代器类型：
//! - ArrayIterator 逐元素遍历数组
//! - StringIterator 按 Unicode 码点遍历字符串
//! - RangeIterator 遍历整数区间

const std = @import("std");
const obj_header = @import("obj_header.zig");
const ObjHeader = obj_header.ObjHeader;
const ThreadContext = obj_header.ThreadContext;
const value = @import("mod.zig");
const Value = value.Value;

/// 数组迭代器，持有序列视图和当前索引
pub const ArrayIterator = struct {
    header: ObjHeader = .{ .type_tag = .array_iter },
    array: []Value,
    index: usize,

    /// 空实现：数组引用由外部管理，迭代器本身无需释放
    pub fn deinit(self: *ArrayIterator, tctx: *ThreadContext) void {
        _ = self;
        _ = tctx;
    }
};

/// 字符串迭代器，按字节偏移遍历 UTF-8 编码的字符串
pub const StringIterator = struct {
    header: ObjHeader = .{ .type_tag = .string_iter },
    string: []const u8,
    byte_offset: usize,

    /// 空实现：字符串引用由外部管理，迭代器本身无需释放
    pub fn deinit(self: *StringIterator, tctx: *ThreadContext) void {
        _ = self;
        _ = tctx;
    }
};

/// 区间迭代器，从 current 遍历到 end
pub const RangeIterator = struct {
    header: ObjHeader = .{ .type_tag = .range_iter },
    current: [16]u8,
    end: [16]u8,
    inclusive: bool,

    /// 空实现：整数值为内联类型，迭代器本身无需释放
    pub fn deinit(self: *RangeIterator, tctx: *ThreadContext) void {
        _ = self;
        _ = tctx;
    }
};

// ── deinit_table 注册函数 ──

pub fn arrayIterDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *ArrayIterator = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    tctx.freeObj(@ptrCast(self));
}

pub fn stringIterDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *StringIterator = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    tctx.freeObj(@ptrCast(self));
}

pub fn rangeIterDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *RangeIterator = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    tctx.freeObj(@ptrCast(self));
}

/// 注册所有迭代器类型的 deinit 函数
pub fn registerDeinits() void {
    obj_header.registerDeinit(.array_iter, arrayIterDeinit);
    obj_header.registerDeinit(.string_iter, stringIterDeinit);
    obj_header.registerDeinit(.range_iter, rangeIterDeinit);
}

File: rust/src/value/iterator.rs
//! 迭代器堆对象类型：数组迭代器、字符串迭代器、范围迭代器
//!
//! 对应 Zig 实现中的迭代器值类型。迭代器持有底层数据的引用（`Rc`），
//! 每次 `next` 调用推进内部游标。

use std::rc::Rc;

use crate::value::Value;

// =========================================================================
// ArrayIterator — 数组迭代器
// =========================================================================

/// 数组迭代器：按顺序遍历数组元素
#[derive(Debug, Clone)]
pub struct ArrayIterator {
    /// 底层数组（共享引用）
    pub array: Rc<Vec<Value>>,
    /// 当前索引
    pub index: usize,
}

impl ArrayIterator {
    /// 创建数组迭代器
    pub fn new(array: Rc<Vec<Value>>) -> Self {
        Self { array, index: 0 }
    }

    /// 推进迭代器，返回下一个元素的克隆
    pub fn next(&mut self) -> Option<Value> {
        if self.index < self.array.len() {
            let val = self.array[self.index].clone();
            self.index += 1;
            Some(val)
        } else {
            None
        }
    }
}

// =========================================================================
// StringIterator — 字符串迭代器
// =========================================================================

/// 字符串迭代器：按 Unicode 码点遍历字符串
#[derive(Debug, Clone)]
pub struct StringIterator {
    /// 底层字符串（共享引用）
    pub string: Rc<str>,
    /// 当前字节偏移
    pub byte_offset: usize,
}

impl StringIterator {
    /// 创建字符串迭代器
    pub fn new(string: Rc<str>) -> Self {
        Self { string, byte_offset: 0 }
    }

    /// 推进迭代器，返回下一个码点
    pub fn next(&mut self) -> Option<u32> {
        if self.byte_offset >= self.string.len() {
            return None;
        }
        let rest = &self.string[self.byte_offset..];
        let c = rest.chars().next()?;
        self.byte_offset += c.len_utf8();
        Some(c as u32)
    }
}

// =========================================================================
// RangeIterator — 范围迭代器
// =========================================================================

/// 范围迭代器：按顺序产生范围内的整数
#[derive(Debug, Clone)]
pub struct RangeIterator {
    /// 当前值
    pub current: i64,
    /// 结束值
    pub end: i64,
    /// 是否包含结束值
    pub inclusive: bool,
}

impl RangeIterator {
    /// 创建范围迭代器
    pub fn new(start: i64, end: i64, inclusive: bool) -> Self {
        Self {
            current: start,
            end,
            inclusive,
        }
    }

    /// 推进迭代器，返回下一个整数
    pub fn next(&mut self) -> Option<i64> {
        if self.inclusive {
            if self.current <= self.end {
                let v = self.current;
                self.current += 1;
                Some(v)
            } else {
                None
            }
        } else if self.current < self.end {
            let v = self.current;
            self.current += 1;
            Some(v)
        } else {
            None
        }
    }
}
